src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,395 @@
+            ImmutableArray<IParameterSymbol> parameters)
+        {
+            var formatArgumentSyntax = TryGetArgument(
+                semanticModel,
+                NameOfFormatStringParameter,
+                arguments,
+                parameters);
+
+            if (formatArgumentSyntax == null)
+            {
+                return null;
+            }
+
+            if (!ArgumentExpressionIsStringLiteral(formatArgumentSyntax))
+            {
+                return null;
+            }
+
+            return GetArgumentExpression(formatArgumentSyntax);
+        }
+
+        protected static IMethodSymbol TryGetValidFormatMethodSymbol(
+            SyntaxNodeAnalysisContext context,
+            int numberOfArguments,
+            SymbolInfo symbolInfo)
+        {
+
+            if (symbolInfo.Symbol == null)
+            {
+                return null;
+            }
+
+            if (symbolInfo.Symbol.Kind != SymbolKind.Method)
+            {
+                return null;
+            }
+
+            var containingType = (INamedTypeSymbol)symbolInfo.Symbol.ContainingSymbol;
+
+            if (containingType.SpecialType != SpecialType.System_String)
+            {
+                return null;
+            }
+
+            var method = (IMethodSymbol)symbolInfo.Symbol;
+
+            if (method.Arity > 0)
+            {
+                return null;
+            }
+
+            return method;
+        }
+
+        private bool FormatCallWorksAtRuntime(string formatString, int numberOfPlaceholderArguments)
+        {
+            var testArray = new object[numberOfPlaceholderArguments];
+            for (var i = 0; i < numberOfPlaceholderArguments; i++)
+            {
+                testArray[i] = "test";
+            }
+
+            try
+            {
+                string.Format(formatString, testArray);
+            }
+            catch
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        protected void ValidateAndReportDiagnostic(
+            SyntaxNodeAnalysisContext context,
+            int numberOfPlaceholderArguments,
+            string formatString,
+            int formatStringPosition)
+        {
+            var formatStringWithEscapedBracketsRemoved = RemoveEscapedBrackets(formatString);
+
+            var matches = s_extractPlaceholdersRegex.Matches(formatStringWithEscapedBracketsRemoved);
+            foreach (Match match in matches)
+            {
+                var textInsideBrackets = match.Groups[1].Value;
+
+                if (!PlaceholderIndexIsValid(textInsideBrackets, numberOfPlaceholderArguments))
+                {
+                    var invalidPlaceholderText = "{" + textInsideBrackets + "}";
+                    var invalidPlaceholderLocation = Location.Create(
+                        context.Node.SyntaxTree,
+                        new Text.TextSpan(
+                            formatStringPosition + match.Index, 
```

Comment:
you're making against formatStringWithEscapedBracketsRemoved, but then using indices to position the location inside formatString.  it seems like these locations may be incorrect.  For example, if i had ```"{{    {{   {{   {{   {1}", 0```, will the error message be in the right location?